pkg/utils/scheduling/scheduling.go

```diff
@@ -0,0 +1,107 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package scheduling
+
+import (
+	"go.uber.org/zap"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+)
+
+var (
+	IgnoredOwners []schema.GroupVersionKind = []schema.GroupVersionKind{
+		{Group: "apps", Version: "v1", Kind: "DaemonSet"},
+	}
+)
+
+func IsNotIgnored(pod *v1.Pod) bool {
+	for _, ignoredOwner := range IgnoredOwners {
+		for _, owner := range pod.ObjectMeta.OwnerReferences {
+			if owner.APIVersion == ignoredOwner.GroupVersion().String() && owner.Kind == ignoredOwner.Kind {
+				zap.S().Debugf("Ignoring %s %s %s/%s", owner.APIVersion, owner.Kind, pod.Namespace, owner.Name)
+				return false
+			}
+		}
+	}
+	return true
+}
+
+func IsUnschedulable(pod *v1.Pod) bool {
+	for _, condition := range pod.Status.Conditions {
+		if condition.Type == v1.PodScheduled && condition.Reason == v1.PodReasonUnschedulable {
+			return true
+		}
+	}
+	return false
+}
+
+// GetResources returns the total resources of the pod.
+func GetResources(pod v1.PodSpec) v1.ResourceList {
+	resources := v1.ResourceList{v1.ResourceCPU: resource.Quantity{}, v1.ResourceMemory: resource.Quantity{}}
+	for _, container := range pod.Containers {
+		if cpu := container.Resources.Requests.Cpu(); cpu != nil {
+			resources.Cpu().Add(*cpu)
+		}
+		if memory := container.Resources.Requests.Memory(); memory != nil {
+			resources.Memory().Add(*memory)
+		}
+	}
+	return resources
+}
+
+// IsSchedulable returns true if the pod can schedule to the node
+func IsSchedulable(pod *v1.PodSpec, node *v1.Node) bool {
+	// Tolerate Taints
+	if !ToleratesAllTaints(pod, node.Spec.Taints) {
+		return false
+	}
+	// Match Node Selector labels
+	if !labels.SelectorFromSet(pod.NodeSelector).Matches(labels.Set(node.Labels)) {
+		return false
+	}
+	// TODO, support node affinity
+	return true
+}
+
+// ToleratesAllTaints returns true if the pod tolerates all taints
+func ToleratesAllTaints(pod *v1.PodSpec, taints []v1.Taint) bool {
+	for _, taint := range taints {
+		if !ToleratesTaint(pod, taint) {
+			return false
+		}
+	}
+	return true
+}
+
+// ToleratesTaint returns true if the pod tolerates the taint
+// https://kubernetes.io/docs/concepts/scheduling-eviction/taint-and-toleration/#concepts
+func ToleratesTaint(pod *v1.PodSpec, taint v1.Taint) bool {
+	if taint.Effect == v1.TaintEffectPreferNoSchedule {
+		return true
+	}
+	for _, toleration := range pod.Tolerations {
+		if toleration.Key == taint.Key {
+			if toleration.Operator == v1.TolerationOpExists {
+				return true
+			}
+			if toleration.Operator == v1.TolerationOpEqual && toleration.Value == taint.Value {
+				return true
+			}
+		}
```

Comment:
feel free to ignore - I'd write this with one `if` stmt

Reply:
Keeping as is. I'm not a huge fan of "run on predicates". As soon as there's some question of order of operations of && and ||, I break it apart.